spec/lang/step/operators.md

```diff
@@ -15,22 +15,36 @@ impl<M: Memory> Machine<M> {
 
 ```rust
 impl<M: Memory> Machine<M> {
-    fn eval_int_un_op(op: IntUnOp, operand: Int) -> Result<Int> {
+    fn eval_int_un_op(op: IntUnOp, operand: Int, int_ty: IntType) -> Result<Int> {
         use IntUnOp::*;
         ret(match op {
             Neg => -operand,
             BitNot => !operand,
+            CountOnes => {
+                let mut ones = Int::ZERO;
+                let mut remaining_bits = operand;
+                for _ in Int::ZERO..int_ty.size.bits() {
+                    ones += remaining_bits & Int::ONE;
+                    remaining_bits >>= 1;
+                }
+                ones
+            }
         })
     }
     fn eval_un_op(&self, UnOp::Int(op): UnOp, (operand, op_ty): (Value<M>, Type)) -> Result<(Value<M>, Type)> {
         let Type::Int(int_ty) = op_ty else { panic!("non-integer input to integer operation") };
         let Value::Int(operand) = operand else { panic!("non-integer input to integer operation") };
 
+        let ret_ty = match op {
+            IntUnOp::CountOnes => IntType { signed: Unsigned, size: Size::from_bytes(4).unwrap() },
+            _ => int_ty,
+        };
+
         // Perform the operation.
-        let result = Self::eval_int_un_op(op, operand)?;
+        let result = Self::eval_int_un_op(op, operand, int_ty)?;
         // Put the result into the right range (in case of overflow).
-        let result = int_ty.bring_in_bounds(result);
-        ret((Value::Int(result), Type::Int(int_ty)))
+        let result = ret_ty.bring_in_bounds(result);
```

Comment:
I don't think we want to call `bring_in_bounds` on the result of `CountOnes`. That should be a bug if we ever lose precision there.